lib/redis_memo.rb

```diff
@@ -94,6 +96,29 @@ def self.without_memo
     Thread.current[THREAD_KEY_WITHOUT_MEMO] = prev_value
   end
 
+  def self.with_max_connection_attempts(max_attempts)
+    prev_value = Thread.current[THREAD_KEY_WITHOUT_MEMO]
+    if max_attempts
+      Thread.current[THREAD_KEY_CONNECTION_ATTEMPTS_COUNT] = 0
+      Thread.current[THREAD_KEY_MAX_CONNECTION_ATTEMPTS] = max_attempts
+    end
+    yield
+  ensure
+    Thread.current[THREAD_KEY_WITHOUT_MEMO] = prev_value
+    Thread.current[THREAD_KEY_CONNECTION_ATTEMPTS_COUNT] = nil
+    Thread.current[THREAD_KEY_MAX_CONNECTION_ATTEMPTS] = nil
+  end
+
+  # Fall back to RedisMemo.without_memo if the maximum connection attempts has been reached for a single request
+  def self.incr_max_connection_attempts
+    return if Thread.current[THREAD_KEY_MAX_CONNECTION_ATTEMPTS].nil? || Thread.current[THREAD_KEY_CONNECTION_ATTEMPTS_COUNT].nil?
+
+    Thread.current[THREAD_KEY_CONNECTION_ATTEMPTS_COUNT] += 1
+    if Thread.current[THREAD_KEY_CONNECTION_ATTEMPTS_COUNT] >= Thread.current[THREAD_KEY_MAX_CONNECTION_ATTEMPTS]
+      Thread.current[THREAD_KEY_WITHOUT_MEMO] = true
```

Comment:
maybe let's add a comment explaining this is reset in with_max_connection_attempts